File: Hazelcast.Net/Hazelcast.Client.Protocol.Codec/BuiltIn/CacheSimpleEntryListenerConfigCodec.java
/*
 * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hazelcast.client.impl.protocol.codec.builtin;

import com.hazelcast.client.impl.protocol.ClientMessage;
import com.hazelcast.config.CacheSimpleEntryListenerConfig;
import com.hazelcast.nio.Bits;

import java.util.ListIterator;

import static com.hazelcast.client.impl.protocol.ClientMessage.BeginFrame;
import static com.hazelcast.client.impl.protocol.ClientMessage.EndFrame;
import static com.hazelcast.client.impl.protocol.codec.builtin.CodecUtil.DecodeNullable;
import static com.hazelcast.client.impl.protocol.codec.builtin.CodecUtil.EncodeNullable;
import static com.hazelcast.client.impl.protocol.codec.builtin.CodecUtil.fastForwardToEndFrame;
import static com.hazelcast.client.impl.protocol.codec.builtin.FixedSizeTypesCodec.DecodeBoolean;
import static com.hazelcast.client.impl.protocol.codec.builtin.FixedSizeTypesCodec.EncodeBoolean;

public final class CacheSimpleEntryListenerConfigCodec {
    private const int OLD_VALUE_REQUIRED_OFFSET = 0;
    private const int SYNCHRONOUS_OFFSET = OLD_VALUE_REQUIRED_OFFSET + Bits.BOOLEAN_SIZE_IN_BYTES;
    private const int InitialFrameSize = SYNCHRONOUS_OFFSET + Bits.BOOLEAN_SIZE_IN_BYTES;

    private CacheSimpleEntryListenerConfigCodec() {
    }

    public static void Encode(ClientMessage clientMessage, CacheSimpleEntryListenerConfig config) {
        clientMessage.add(BeginFrame);

        ClientMessage.Frame initialFrame = new ClientMessage.Frame(new byte[InitialFrameSize]);
        EncodeBoolean(initialFrame.content, OLD_VALUE_REQUIRED_OFFSET, config.isOldValueRequired());
        EncodeBoolean(initialFrame.content, SYNCHRONOUS_OFFSET, config.isSynchronous());
        clientMessage.add(initialFrame);

        EncodeNullable(clientMessage, config.getCacheEntryListenerFactory(), StringCodec::Encode);
        EncodeNullable(clientMessage, config.getCacheEntryEventFilterFactory(), StringCodec::Encode);

        clientMessage.add(EndFrame);
    }

    public static CacheSimpleEntryListenerConfig Decode(ListIterator<ClientMessage.Frame> iterator) {
        // begin frame
        iterator.next();

        ClientMessage.Frame initialFrame = iterator.next();
        boolean oldValueRequired = DecodeBoolean(initialFrame.content, OLD_VALUE_REQUIRED_OFFSET);
        boolean synchronous = DecodeBoolean(initialFrame.content, SYNCHRONOUS_OFFSET);

        String cacheEntryListenerFactory = DecodeNullable(iterator, StringCodec::Decode);
        String cacheEntryEventFilterFactory = DecodeNullable(iterator, StringCodec::Decode);

        fastForwardToEndFrame(iterator);

        CacheSimpleEntryListenerConfig config = new CacheSimpleEntryListenerConfig();
        config.setOldValueRequired(oldValueRequired);
        config.setSynchronous(synchronous);
        config.setCacheEntryListenerFactory(cacheEntryListenerFactory);
        config.setCacheEntryEventFilterFactory(cacheEntryEventFilterFactory);
        return config;
    }
}

File: Hazelcast.Net/Hazelcast.Client.Protocol.Codec/custom/CacheSimpleEntryListenerConfigCodec.cs
// Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
// http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections;
using System.Collections.Generic;
using Hazelcast.Client.Protocol;
using Hazelcast.Client.Protocol.Codec.BuiltIn;
using Hazelcast.Client.Protocol.Util;
using Hazelcast.IO;
using static Hazelcast.Client.Protocol.Codec.BuiltIn.FixedSizeTypesCodec;
using static Hazelcast.Client.Protocol.ClientMessage;
using static Hazelcast.IO.Bits;

namespace Hazelcast.Client.Protocol.Codec.Custom
{
    // This file is auto-generated by the Hazelcast Client Protocol Code Generator.
    // To change this file, edit the templates or the protocol
    // definitions on the https://github.com/hazelcast/hazelcast-client-protocol
    // and regenerate it.

    internal static class CacheSimpleEntryListenerConfigCodec 
    {
        private const int OldValueRequiredFieldOffset = 0;
        private const int SynchronousFieldOffset = OldValueRequiredFieldOffset + BoolSizeInBytes;
        private const int InitialFrameSize = SynchronousFieldOffset + BoolSizeInBytes;

        public static void Encode(ClientMessage clientMessage, com.hazelcast.config.CacheSimpleEntryListenerConfig cacheSimpleEntryListenerConfig) 
        {
            clientMessage.Add(BeginFrame);

            var initialFrame = new Frame(new byte[InitialFrameSize]);
            EncodeBool(initialFrame.Content, OldValueRequiredFieldOffset, cacheSimpleEntryListenerConfig.IsOldValueRequired);
            EncodeBool(initialFrame.Content, SynchronousFieldOffset, cacheSimpleEntryListenerConfig.IsSynchronous);
            clientMessage.Add(initialFrame);

            CodecUtil.EncodeNullable(clientMessage, cacheSimpleEntryListenerConfig.CacheEntryListenerFactory, StringCodec.Encode);
            CodecUtil.EncodeNullable(clientMessage, cacheSimpleEntryListenerConfig.CacheEntryEventFilterFactory, StringCodec.Encode);

            clientMessage.Add(EndFrame);
        }

        public static com.hazelcast.config.CacheSimpleEntryListenerConfig Decode(ref FrameIterator iterator) 
        {
            // begin frame
            iterator.Next();

            ref var initialFrame = ref iterator.Next();
            var oldValueRequired = DecodeBool(initialFrame.Content, OldValueRequiredFieldOffset);
            var synchronous = DecodeBool(initialFrame.Content, SynchronousFieldOffset);

            var cacheEntryListenerFactory = CodecUtil.DecodeNullable(ref iterator, StringCodec.Decode);
            var cacheEntryEventFilterFactory = CodecUtil.DecodeNullable(ref iterator, StringCodec.Decode);

            CodecUtil.FastForwardToEndFrame(ref iterator);

            return CustomTypeFactory.CreateCacheSimpleEntryListenerConfig(oldValueRequired, synchronous, cacheEntryListenerFactory, cacheEntryEventFilterFactory);
        }
    }
}